x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -649,142 +711,270 @@ private void indexInvalidation(Collection<String> tokenIds, ActionListener<Token
-                                            }
-                                        }),
-                                    client::update);
-                            }
+                            logger.warn("could not find superseding token document [{}] for token document [{}] after all retries",
+                                supersedingTokenDocId, tokenDocId);
+                            onFailure.accept(invalidGrantException("could not refresh the requested token"));
                         }
-                    } else {
-                        logger.info("could not find token document [{}] for refresh", tokenDocId);
-                        onFailure.accept(invalidGrantException("could not refresh the requested token"));
                     }
-                }, e -> {
-                    if (isShardNotAvailableException(e)) {
-                        innerRefresh(tokenDocId, clientAuth, listener, attemptCount);
-                    } else {
-                        listener.onFailure(e);
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        if (isShardNotAvailableException(e)) {
+                            if (backoff.hasNext()) {
+                                logger.info("could not find superseding token document [{}] for refresh, retrying", supersedingTokenDocId);
+                                client.threadPool().schedule(
+                                    () -> getTokenDocAsync(supersedingTokenDocId, this), backoff.next(), GENERIC);
+                            } else {
+                                logger.warn("could not find token document [{}] for refresh after all retries", supersedingTokenDocId);
+                                onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                            }
+                        } else {
+                            logger.warn("could not find superseding token document [{}] for refresh", supersedingTokenDocId);
+                            onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                        }
                     }
-                }), client::get);
+                };
+                getTokenDocAsync(supersedingTokenDocId, getSupersedingListener);
+            } else {
+                final Map<String, Object> userTokenSource = (Map<String, Object>)
+                    ((Map<String, Object>) source.get("access_token")).get("user_token");
+                final String authString = (String) userTokenSource.get("authentication");
+                final Integer version = (Integer) userTokenSource.get("version");
+                final Map<String, Object> metadata = (Map<String, Object>) userTokenSource.get("metadata");
+                Version authVersion = Version.fromId(version);
+                try (StreamInput in = StreamInput.wrap(Base64.getDecoder().decode(authString))) {
```

Comment:
nit: this try/catch extends as much as possible instead of as little as possible.